### PR TITLE
ICMSLST-1947 

### PR DESCRIPTION
**Pull request by MattHolmes123**
Changes:
  - Move all workbasket orm code to app_data.py.
  - Get querysets based on importer / exporter access permission to reduces the
    number of db queries.
  - Fix duplicate queries relating to issued documents.
  - Add ActionConfig to share data between all actions to fix duplicate queries.
  - Fix several workbasket actions:
    - Only show Issued Document actions if the user can edit the app.
    - Only show Firearm Report action if the user can edit the app.
    - Show Manage action when an application update has been sent.
    - Add more information to section label for FIR and update requests.
  - Remove task checking from applicant FIR views.
    Previously if an Update Request and FIR were sent at the same time it would
    cause the FIR views to throw an exception.
  - Several fixes for update request views:
    - Allow ILB Admin to manage case when there is an open update request
    - Fix template bugs when an update request is closed with no applicant
      response (when opened in error).
  - Several fixes for caseworker workbasket:
    - No longer show complete applications.
    - No longer show revoked applications.
  - Remove several Action attributes:
    - is_rejected
    - is_importer_user
    - is_ilb_admin

### CODE REVIEW

File: web/domains/workbasket/actions/ilb_admin_actions.py [hunk 1]
```diff
@@ -172,12 +174,21 @@ def show_link(self) -> bool:
 
     def get_workbasket_actions(self) -> list[WorkbasketAction]:
         kwargs = self.get_kwargs()
+
+        section_label = "Application Processing"
```
Comment (marcuspp) [inline web/domains/workbasket/actions/ilb_admin_actions.py:178]: This is the same as self.section_label

File: web/domains/workbasket/actions/base.py [hunk 2]
```diff
@@ -15,22 +32,28 @@ class Action:
...
         self.application = application
         self.active_tasks = tasks
         self.status = self.application.status
-        self.is_ilb_admin = is_ilb_admin
-        self.is_importer_user = is_importer_user
-        self.is_rejected = is_rejected
+        self.app_checker = app_checker
+
+    @classmethod
+    def from_config(cls, config: ActionConfig):
```
Comment (marcuspp) [inline web/domains/workbasket/actions/base.py:49]: Missing return type

File: web/domains/workbasket/actions/base.py [hunk 3]
```diff
@@ -44,5 +67,17 @@ def is_case_owner(self) -> bool:
     def get_kwargs(self) -> dict[str, Any]:
         return {"application_pk": self.application.pk, "case_type": self.case_type}
 
+    def has_task(self, *tasks):
```
Comment (marcuspp) [inline web/domains/workbasket/actions/base.py:70]: Missing return type

File: web/domains/workbasket/app_data.py [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+        # Has permission at the agent directly (agent contacts)
+        (Q(agent__in=agent_exporters))
+    )
+
+    return [exporter_approval_requests, export_applications]
+
+
+def _add_user_import_annotations(
+    applications: "QuerySet[ImportApplication]",
+) -> "QuerySet[ImportApplication]":
```
Comment (marcuspp) [inline web/domains/workbasket/app_data.py:348]: Types don't need to be strings as already imported

File: web/domains/workbasket/actions/applicant_actions.py [hunk 5]
```diff
@@ -108,6 +105,22 @@ def get_workbasket_actions(self) -> list[WorkbasketAction]:
...
+
+        if self.status in [ImpExpStatus.COMPLETED, ImpExpStatus.REVOKED]:
+            show_link = True
+
+        return show_link
+
+    def get_workbasket_actions(self) -> list[WorkbasketAction]:
+        section_label = "Application View"
+
+        # ICMSLST-19 Add clear action
```
Comment (marcuspp) [inline web/domains/workbasket/actions/applicant_actions.py:120]: Does a ticket need to be created for this?

File: web/domains/workbasket/actions/base.py [hunk 6]
```diff
@@ -1,10 +1,27 @@
...
+class ActionConfig:
+    """Cache shared between all actions."""
+
+    user: User
+    case_type: str
+    application: ImpOrExp
+    tasks: list[str] = dataclasses.field(init=False)
+    app_checker: AppChecker = dataclasses.field(init=False)
+
+    def __post_init__(self):
```
Comment (marcuspp) [inline web/domains/workbasket/actions/base.py:20]: Missing return type

File: web/domains/workbasket/actions/utils.py [hunk 7]
```diff
@@ -1,64 +1,40 @@
...
-
-    from .base import ActionT
+from .applicant_actions import APPLICANT_ACTIONS
+from .base import ActionConfig, ActionT
+from .ilb_admin_actions import CASEWORKER_ACTIONS
 
 
 def get_workbasket_admin_sections(
-    user: "User", case_type: str, application: "ImpOrExp"
+    user: "User", case_type: str, application: ImpOrExp
```
Comment (marcuspp) [inline web/domains/workbasket/actions/utils.py:13]: "User" can be User

**Review by marcuspp — APPROVED** [review]

File: web/domains/workbasket/actions/applicant_actions.py (hunk 5 above)
Comment (MattHolmes123) [reply web/domains/workbasket/actions/applicant_actions.py:120]: No "ICMSLST-19" is the ticket :)

File: web/domains/workbasket/actions/base.py (hunk 2 above)
Comment (MattHolmes123) [reply web/domains/workbasket/actions/base.py:49]: Sorted.

File: web/domains/workbasket/actions/base.py (hunk 3 above)
Comment (MattHolmes123) [reply web/domains/workbasket/actions/base.py:70]: fixed

File: web/domains/workbasket/actions/utils.py (hunk 7 above)
Comment (MattHolmes123) [reply web/domains/workbasket/actions/utils.py:13]: fixed

File: web/domains/workbasket/app_data.py [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+        .annotate(annotation_open_fir_pks=open_fir_pks_annotation)
+        .filter(
+            is_active=True,
+            status=ApprovalRequest.OPEN,
+            access_request__importeraccessrequest__link__in=main_importers,
+        )
+    )
+
+    # Import Applications
+    import_applications = ImportApplication.objects.all().select_related(
```
Comment (mikeylockwood) [inline web/domains/workbasket/app_data.py:256]: There's always an `.all()` to find :)

File: web/domains/workbasket/app_data.py [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+        .annotate(annotation_open_fir_pks=open_fir_pks_annotation)
+        .filter(
+            is_active=True,
+            status=ApprovalRequest.OPEN,
+            access_request__exporteraccessrequest__link__in=main_exporters,
+        )
+    )
+
+    # Export Applications
+    export_applications = ExportApplication.objects.all().select_related(
```
Comment (mikeylockwood) [inline web/domains/workbasket/app_data.py:320]: Another `.all()`

**Pull request merged**